app/views/submissions/_annotation_pane.html.erb

```diff
@@ -3,48 +3,101 @@
-            </h4>
-            <i class="large material-icons expand-icon">arrow_drop_down</i>
-            <i class="large material-icons collapse-icon">arrow_drop_up</i>
+          <div class="collapsible-header-wrap">
+            <div class="collapsible-header">
+              <h4> <%= problem.capitalize %>
+                  <div class="summary_score">
+                    <% if @cud.instructor? or @cud.course_assistant? or @problemReleased %>
+                      <%= plus_fix(@problemGrades[problem]) %>
+                    <% end %>
+                  </div>
+              </h4>
+            </div>
+            <div class="collapsible-header-controls">
+              <a href="#" title="Expand dropdown"><i class="small material-icons expand-icon">expand_more</i></a>
+              <a href="#" title="Collapse dropdown"><i class="small material-icons collapse-icon">expand_less</i></a>
+              <% if @cud.instructor? or @cud.course_assistant? %>
+              <a href="#" class="global-annotation-add-button" data-problem="<%= problem %>" title="Add global annotation">
+                <i class="small material-icons">add</i>
+              </a>
+              <% end %>
+            </div>
           </div>
           <div class="collapsible-body">
-            <% if @cud.instructor? or @cud.course_assistant? or @problemReleased%>
-              <!-- a[6] = filename, a[2] = line-->
+            <% if @cud.instructor? or @cud.course_assistant? or @problemReleased %>
+              <% if descriptTuples.empty? %>
+                There are currently no annotations.
+              <% end %>
               <%# Note: group_by preserves insertion order %>
-              <% annotations_by_file = descriptTuples.sort_by{ |a| [a[6], a[2]] }.group_by { |a| a[6] }
-                 annotations_by_file.each do |filename, annotations| %>
-                <div class="file_annotations">
-                  <strong> <%= filename %> </strong>
-                  <% annotations.each do |description, value, line, user, id, position, filename| %>
-                    <div class="descript" id="li-annotation-<%= id %>"> <!-- Line + 1 because the code line numbers starts with 1 not 0. -->
-                      <%= link_to(
-                          url_for([:view, @course, @assessment, @submission, header_position: position.nil? ? 0 : position, line: line.nil? ? 1 : line+1]),
-                          class: 'descript-link',
-                          "data-header_position": position.nil? ? 0 : position,
-                          "data-line": line.nil? ? 1 : line+1,
-                          remote: true,
-                        ) do %>
+              <%#
+                  First, group_by global (a[7])
+                  Sort global annotations by id (a[4])
+                  Sort file annotations by filename (a[6]), followed by line (a[2]) and group by filename
+              %>
+              <% annotations_by_type = descriptTuples.group_by { |a| a[7] }
```

Comment:
would it be possible to move this logic to the controller?

Reply:
Makes sense